File: impl/csharp/test/StreamingTests.cs
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

using Naalp;
using Xunit;

namespace Sh.Bubblefish.Naalp.Tests
{
    /// <summary>
    /// C9 — native streaming with a single signed per-stream commitment for the C# SDK (design.md §10;
    /// R-10.1..10.6), ported from impl/go/streaming and cross-checked against
    /// impl/python/naalp/streaming.py. Graded against the shared independent corpus
    /// <c>vectors/stream/cases.json</c> (the oracle directory is <c>stream</c>, NOT <c>streaming</c>),
    /// which is NOT produced by this code.
    ///
    /// <para>BYTE surface (⟹ csharp == Go == Rust == Python == oracle): the StreamOpen / StreamCommit /
    /// StreamCheckpoint body hexes, the final rolling-SHA-384 commitment, each checkpoint digest_so_far,
    /// and the oracle-pinned tampered digest. Behaviour: order-independence of the commitment
    /// (reversed-input-same-digest) versus offset-sensitivity (swapped-offset-different-digest), prefix
    /// checkpoints, and tamper → StreamDigestMismatch. FULL-SIG (real deterministic ML-DSA-65, rnd=0):
    /// the raw StreamOpen/Commit/Checkpoint signatures round-trip (a raw ML-DSA signature over the body,
    /// matching impl/go/streaming; not corpus-graded).</para>
    /// </summary>
    public sealed class StreamingTests
    {
        private const int Alg = Cose.ALG_MLDSA65;

        private static JsonElement Vector()
        {
            var d = new DirectoryInfo(AppContext.BaseDirectory);
            for (int i = 0; i < 12 && d != null; i++)
            {
                string p = Path.Combine(d.FullName, "vectors", "stream", "cases.json");
                if (File.Exists(p))
                {
                    return JsonDocument.Parse(File.ReadAllText(p)).RootElement.Clone();
                }
                d = d.Parent;
            }
            throw new Xunit.Sdk.XunitException("vectors/stream/cases.json not found from " + AppContext.BaseDirectory);
        }

        private static string Hex(byte[] b) => Convert.ToHexString(b).ToLowerInvariant();
        private static byte[] Hb(string s) => Convert.FromHexString(s);

        private static List<Streaming.Chunk> Chunks(JsonElement v)
        {
            var outp = new List<Streaming.Chunk>();
            foreach (JsonElement c in v.GetProperty("chunks").EnumerateArray())
            {
                outp.Add(new Streaming.Chunk(c.GetProperty("offset").GetInt64(), Hb(c.GetProperty("data_hex").GetString()!)));
            }
            return outp;
        }

        // ---- StreamOpen / StreamCommit / StreamCheckpoint body parity + the rolling commitment ------

        [Fact]
        public void BodyParityAndCommitment()
        {
            JsonElement v = Vector();
            byte[] streamId = Hb(v.GetProperty("stream_id_hex").GetString()!);
            long substream = v.GetProperty("substream").GetInt64();
            long effect = v.GetProperty("effect").GetInt64();
            byte[] approval = Hb(v.GetProperty("approval_hex").GetString()!);

            var open = new Streaming.StreamOpen(streamId, effect, approval, substream);
            Assert.Equal(v.GetProperty("open_body_hex").GetString(), Hex(open.Bytes()));

            List<Streaming.Chunk> chunks = Chunks(v);
            byte[] digest = Streaming.CommitDigest(chunks);
            Assert.Equal(v.GetProperty("final_digest_hex").GetString(), Hex(digest));

            var commit = new Streaming.StreamCommit(streamId, digest);
            Assert.Equal(v.GetProperty("commit_body_hex").GetString(), Hex(commit.Bytes()));

            // The first checkpoint's body encodes byte-identically (through_offset 6, digest_so_far).
            JsonElement cp0 = v.GetProperty("checkpoints")[0];
            var checkpoint = new Streaming.StreamCheckpoint(streamId, cp0.GetProperty("through_offset").GetInt64(),
                Hb(cp0.GetProperty("digest_so_far_hex").GetString()!));
            Assert.Equal(v.GetProperty("checkpoint_body_hex").GetString(), Hex(checkpoint.Bytes()));

            // Every checkpoint's digest_so_far is the rolling digest of the prefix through its offset.
            foreach (JsonElement cp in v.GetProperty("checkpoints").EnumerateArray())
            {
                long through = cp.GetProperty("through_offset").GetInt64();
                var prefix = new List<Streaming.Chunk>();
                foreach (Streaming.Chunk c in chunks) if (c.Offset < through) prefix.Add(c);
                Assert.Equal(cp.GetProperty("digest_so_far_hex").GetString(), Hex(Streaming.CommitDigest(prefix)));
                // VerifyCheckpoint confirms the prefix.
                var checkpointObj = new Streaming.StreamCheckpoint(streamId, through, Streaming.CommitDigest(prefix));
                Streaming.VerifyCheckpoint(checkpointObj, prefix); // must not throw
            }

            // VerifyCommit confirms the completed commitment.
            Streaming.VerifyCommit(commit, chunks); // must not throw
        }

        // ---- order-independence of the commitment (MUTATION ANCHOR) --------------------------------
        // MUTATION ANCHOR: dropping the absolute-offset sort in CommitDigest makes a reversed delivery
        // of the SAME frames fold in a different order, flipping the reversed-input-same-digest
        // Assert.Equal below.

        [Fact]
        public void CommitmentIsOrderIndependentButOffsetSensitive()
        {
            JsonElement v = Vector();
            List<Streaming.Chunk> chunks = Chunks(v);
            byte[] forward = Streaming.CommitDigest(chunks);

            // A reversed delivery of the SAME frames (same offset->data pairing) yields the SAME digest:
            // CommitDigest folds in absolute-offset order regardless of delivery order.
            var reversed = new List<Streaming.Chunk>(chunks);
            reversed.Reverse();
            Assert.Equal(Hex(forward), Hex(Streaming.CommitDigest(reversed)));
            Assert.Equal(v.GetProperty("final_digest_hex").GetString(), Hex(Streaming.CommitDigest(reversed)));

            // But SWAPPING which bytes sit at which offset changes the digest (offset-sensitive): the
            // commitment is over data-in-offset-order, not a set of frames.
            var swapped = new List<Streaming.Chunk>
            {
                new Streaming.Chunk(chunks[0].Offset, chunks[1].Data),
                new Streaming.Chunk(chunks[1].Offset, chunks[0].Data),
                chunks[2],
            };
            Assert.NotEqual(Hex(forward), Hex(Streaming.CommitDigest(swapped)));
        }

        // ---- tamper: a flipped delivered byte breaks the commitment (oracle-pinned tamper digest) ---

        [Fact]
        public void TamperBreaksCommitment()
        {
            JsonElement v = Vector();
            byte[] streamId = Hb(v.GetProperty("stream_id_hex").GetString()!);
            List<Streaming.Chunk> chunks = Chunks(v);
            byte[] goodDigest = Streaming.CommitDigest(chunks);
            var commit = new Streaming.StreamCommit(streamId, goodDigest);

            JsonElement tamper = v.GetProperty("tamper");
            int idx = tamper.GetProperty("chunk_index").GetInt32();
            var tampered = new List<Streaming.Chunk>(chunks);
            tampered[idx] = new Streaming.Chunk(chunks[idx].Offset, Hb(tamper.GetProperty("flipped_data_hex").GetString()!));

            // The tampered stream's recomputed digest equals the oracle-pinned tampered digest (a
            // non-circular check: the expected value is in the corpus, not from this code).
            Assert.Equal(tamper.GetProperty("digest_hex").GetString(), Hex(Streaming.CommitDigest(tampered)));

            // Verifying the tampered delivery against the honest commitment fails closed.
            var ex = Assert.Throws<NaalpException>(() => Streaming.VerifyCommit(commit, tampered));
            Assert.Equal("StreamDigestMismatch", ex.Kind);

            // A non-contiguous prefix, or a wrong through_offset, is StreamDigestMismatch too.
            var badPrefix = new List<Streaming.Chunk> { chunks[0], chunks[2] }; // gap: offset 6..16 missing
            var cp = new Streaming.StreamCheckpoint(streamId, 6, Streaming.CommitDigest(new List<Streaming.Chunk> { chunks[0] }));
            var exCp = Assert.Throws<NaalpException>(() => Streaming.VerifyCheckpoint(cp, badPrefix));
            Assert.Equal("StreamDigestMismatch", exCp.Kind);
        }

        // ---- effect authorization + full-signature round-trip (real ML-DSA, demonstrated isolation) --

        [Fact]
        public void OpenAuthorizationAndSignatures()
        {
            JsonElement v = Vector();
            byte[] streamId = Hb(v.GetProperty("stream_id_hex").GetString()!);
            long substream = v.GetProperty("substream").GetInt64();
            long effect = v.GetProperty("effect").GetInt64(); // idempotent_write (1)
            byte[] approval = Hb(v.GetProperty("approval_hex").GetString()!);
            var open = new Streaming.StreamOpen(streamId, effect, approval, substream);

            // A ceiling that authorizes idempotent_write admits the stream; a read-only ceiling refuses.
            Streaming.OpenStream(open, Policy.IDEMPOTENT_WRITE); // must not throw
            var ex = Assert.Throws<NaalpException>(() => Streaming.OpenStream(open, Policy.READ_ONLY));
            Assert.Equal("EffectNotAuthorized", ex.Kind);

            byte[] seed = new byte[32];
            for (int i = 0; i < 32; i++) seed[i] = 0x11;
            byte[] pk = Cose.MldsaKeygen("ML-DSA-65", seed);
            byte[] foreignSeed = new byte[32];
            for (int i = 0; i < 32; i++) foreignSeed[i] = 0x22;
            byte[] foreignPk = Cose.MldsaKeygen("ML-DSA-65", foreignSeed);

            List<Streaming.Chunk> chunks = Chunks(v);
            var commit = new Streaming.StreamCommit(streamId, Streaming.CommitDigest(chunks));
            var checkpoint = new Streaming.StreamCheckpoint(streamId,
                v.GetProperty("checkpoints")[0].GetProperty("through_offset").GetInt64(),
                Hb(v.GetProperty("checkpoints")[0].GetProperty("digest_so_far_hex").GetString()!));

            // The three signed objects verify under the owner's key and NOT under a foreign key.
            byte[] openSig = Streaming.SignOpen(open, Alg, seed);
            Assert.True(Streaming.VerifyOpenSig(open, Alg, pk, openSig));
            Assert.False(Streaming.VerifyOpenSig(open, Alg, foreignPk, openSig));

            byte[] commitSig = Streaming.SignCommit(commit, Alg, seed);
            Assert.True(Streaming.VerifyCommitSig(commit, Alg, pk, commitSig));

            byte[] cpSig = Streaming.SignCheckpoint(checkpoint, Alg, seed);
            Assert.True(Streaming.VerifyCheckpointSig(checkpoint, Alg, pk, cpSig));
        }
    }
}

File: impl/kotlin/src/test/kotlin/StreamingTests.kt
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp

import java.io.File
import kotlin.system.exitProcess

/**
 * C9 native-streaming known-answer test for the Kotlin SDK (design.md section 10; R-10.1..10.6),
 * graded against the shared independent corpus vectors/stream/cases.json (the oracle directory is
 * `stream`, NOT `streaming`; the module is named streaming). The corpus was NOT produced by this
 * code.
 *
 * CORPUS-GRADED (pure): the rolling SHA-384 commitment over chunks in absolute-offset order, the
 * mid-stream checkpoint digests, the StreamOpen/StreamCommit/StreamCheckpoint bodies, the tamper
 * digest, and the offset-order-independence property (reversed input yields the same commitment;
 * swapping which bytes sit at which offset changes it). CRYPTO-GRADED (real FIPS-204 ML-DSA-65 via
 * BouncyCastle -- Kotlin is a full-signature port): the single end-commitment signature covering the
 * whole stream verifies, and a tampered signature does not. Full-duplex concurrency (R-10.5, Go
 * -race) is not applicable to a single-thread KAT; the load-bearing offset-order-independence
 * property it protects is graded directly here instead.
 *
 * KAT convention (a standalone main() exiting non-zero on any failure); the filename carries the
 * "Tests" token the ten-language-parity gate indexes. Written test-first: [Streaming] is absent until
 * Streaming.kt lands, so this fails RED with a kotlinc "unresolved reference: Streaming"; a mutation
 * that drops the absolute-offset sort in CommitDigest flips "reversed-input same commitment".
 */

private var stFails = 0

private fun stCheck(name: String, got: String, want: String) {
    if (got == want) {
        println("  ok   $name")
    } else {
        stFails++
        println("  FAIL $name\n       got  $got\n       want $want")
    }
}

// ---- minimal regex JSON access (no JSON library on the Kotlin port) ----

private fun stFindVector(): File {
    var d: File? = File(".").absoluteFile
    repeat(6) {
        val cur = d ?: throw AssertionError("vectors/stream/cases.json not found")
        val p = File(File(cur, "vectors"), "stream/cases.json")
        if (p.isFile) return p
        d = cur.parentFile
    }
    throw AssertionError("vectors/stream/cases.json not found from ${File(".").absolutePath}")
}

/** The string value of a "key": "value" pair inside [scope]. */
private fun stField(scope: String, key: String): String {
    val m = Regex("\"$key\"\\s*:\\s*\"([^\"]*)\"").find(scope)
        ?: throw AssertionError("string key not found: $key")
    return m.groupValues[1]
}

/** The integer value of a "key": <number> pair inside [scope]. */
private fun stIntField(scope: String, key: String): Long {
    val m = Regex("\"$key\"\\s*:\\s*(\\d+)").find(scope)
        ?: throw AssertionError("int key not found: $key")
    return m.groupValues[1].toLong()
}

/** The flat inner { ... } block of the object-valued field named [key] (no nesting). */
private fun stSubObject(json: String, key: String): String {
    val m = Regex("\"$key\"\\s*:\\s*\\{([^{}]*)\\}", RegexOption.DOT_MATCHES_ALL).find(json)
        ?: throw AssertionError("object key not found: $key")
    return m.groupValues[1]
}

/** The flat object blocks of the array named [arrayKey]. */
private fun stObjectBlocks(json: String, arrayKey: String): List<String> {
    val a = Regex("\"$arrayKey\"\\s*:\\s*\\[(.*?)\\]", RegexOption.DOT_MATCHES_ALL).find(json)
        ?: throw AssertionError("array key not found: $arrayKey")
    return Regex("\\{([^{}]*)\\}", RegexOption.DOT_MATCHES_ALL).findAll(a.groupValues[1])
        .map { it.groupValues[1] }.toList()
}

private fun streamingRun() {
    val json = stFindVector().readText(Charsets.UTF_8)

    val streamId = Hex.decode(stField(json, "stream_id_hex"))
    val substream = stIntField(json, "substream")
    val effect = stIntField(json, "effect")
    val approval = Hex.decode(stField(json, "approval_hex"))
    val finalDigest = stField(json, "final_digest_hex")

    val chunks = stObjectBlocks(json, "chunks").map {
        Streaming.Chunk(stIntField(it, "offset"), Hex.decode(stField(it, "data_hex")))
    }
    val checkpoints = stObjectBlocks(json, "checkpoints").map {
        Pair(stIntField(it, "through_offset"), stField(it, "digest_so_far_hex"))
    }

    // 1. the rolling commitment over chunks in absolute-offset order equals the oracle.
    stCheck("commit digest == oracle", Hex.encode(Streaming.commitDigest(chunks)), finalDigest)

    // 2. rolling: DigestSoFar after each chunk equals the corresponding checkpoint; the last equals
    //    the final commitment (the digest is not finalized by a checkpoint, so streaming continues).
    val sd = Streaming.StreamDigest()
    for ((i, ch) in chunks.withIndex()) {
        sd.update(ch.data)
        if (i < checkpoints.size) {
            stCheck("checkpoint $i digest_so_far == oracle", Hex.encode(sd.digestSoFar()), checkpoints[i].second)
        }
    }
    stCheck("rolling final == final_digest", Hex.encode(sd.digestSoFar()), finalDigest)

    // 3. the StreamOpen / StreamCommit / StreamCheckpoint bodies equal the oracle bytes.
    val open = Streaming.StreamOpen(streamId, effect, approval, substream)
    stCheck("StreamOpen body == oracle", Hex.encode(open.bytes()), stField(json, "open_body_hex"))
    val commit = Streaming.StreamCommit(streamId, Hex.decode(finalDigest))
    stCheck("StreamCommit body == oracle", Hex.encode(commit.bytes()), stField(json, "commit_body_hex"))
    val cp0 = Streaming.StreamCheckpoint(streamId, checkpoints[0].first, Hex.decode(checkpoints[0].second))
    stCheck("StreamCheckpoint body == oracle", Hex.encode(cp0.bytes()), stField(json, "checkpoint_body_hex"))

    // 4. R-10.2 tamper: the valid stream verifies; delivering a tampered chunk in place of the
    //    corpus tamper index invalidates the commitment (StreamDigestMismatch). The tampered stream's
    //    recomputed commitment also equals the independent tamper oracle.
    val tamper = stSubObject(json, "tamper")
    val tamperIdx = stIntField(tamper, "chunk_index").toInt()
    stCheck("valid stream verifies", verifyKind { Streaming.verifyCommit(commit, chunks) }, "ok")
    val tampered = chunks.toMutableList()
    tampered[tamperIdx] = Streaming.Chunk(chunks[tamperIdx].offset, Hex.decode(stField(tamper, "flipped_data_hex")))
    stCheck("tampered stream rejected", verifyKind { Streaming.verifyCommit(commit, tampered) }, "StreamDigestMismatch")
    stCheck("tampered commit digest == oracle", Hex.encode(Streaming.commitDigest(tampered)), stField(tamper, "digest_hex"))

    // 5. checkpoints confirm a prefix WITHOUT the end; a prefix of the wrong length is rejected.
    for ((i, c) in checkpoints.withIndex()) {
        val cp = Streaming.StreamCheckpoint(streamId, c.first, Hex.decode(c.second))
        val prefix = chunks.subList(0, i + 1)
        stCheck("checkpoint $i verifies prefix", verifyKind { Streaming.verifyCheckpoint(cp, prefix) }, "ok")
    }
    val cpFull = Streaming.StreamCheckpoint(streamId, checkpoints[0].first, Hex.decode(checkpoints[0].second))
    stCheck("checkpoint rejects wrong-length prefix", verifyKind { Streaming.verifyCheckpoint(cpFull, chunks) }, "StreamDigestMismatch")

    // 6. R-10.3 effect refused before any chunk: a destructive stream under a read-only grant is
    //    refused; the corpus stream (idempotent_write) under an idempotent_write grant is authorized;
    //    an unrecognized effect fails closed to destructive and is refused below destructive.
    val destructive = Streaming.StreamOpen(streamId, Policy.DESTRUCTIVE, null, 1)
    stCheck("destructive under read-only refused", verifyKind { Streaming.openStream(destructive, Policy.READ_ONLY) }, "EffectNotAuthorized")
    val okOpen = Streaming.StreamOpen(streamId, effect, approval, substream)
    stCheck("idempotent stream authorized", verifyKind { Streaming.openStream(okOpen, Policy.IDEMPOTENT_WRITE) }, "ok")
    val unknownEff = Streaming.StreamOpen(streamId, 99, null, 1)
    stCheck("unknown effect refused below destructive", verifyKind { Streaming.openStream(unknownEff, Policy.NON_IDEMPOTENT_WRITE) }, "EffectNotAuthorized")

    // 7. CRYPTO-GRADED (real ML-DSA-65): the ONE end-commitment signature covers the whole stream and
    //    verifies; a tampered signature does not (the single signature makes the stream non-repudiable).
    val seed = ByteArray(32) { 0x3c }
    val pk = Cose.mldsaKeygen("ML-DSA-65", seed)
    val sig = Streaming.signCommit(commit, Cose.ALG_MLDSA65, seed)
    stCheck("commit signature verifies", Streaming.verifyCommitSig(commit, Cose.ALG_MLDSA65, pk, sig).toString(), "true")
    val badSig = sig.copyOf()
    badSig[0] = (badSig[0].toInt() xor 0x01).toByte()
    stCheck("tampered commit signature rejected", Streaming.verifyCommitSig(commit, Cose.ALG_MLDSA65, pk, badSig).toString(), "false")

    // 8. offset-order-independence (the property Go proves with -race + reversed/swapped input): the
    //    commitment folds chunks in ABSOLUTE-OFFSET order, so a reversed delivery of the same frames
    //    yields the SAME commitment, while swapping which bytes sit at which offset changes it.
    val reversed = chunks.reversed()
    stCheck("reversed-input same commitment", Hex.encode(Streaming.commitDigest(reversed)), finalDigest)
    val swapped = chunks.toMutableList()
    swapped[0] = Streaming.Chunk(chunks[0].offset, chunks[1].data)
    swapped[1] = Streaming.Chunk(chunks[1].offset, chunks[0].data)
    stCheck("swapped-offset content different commitment", (Hex.encode(Streaming.commitDigest(swapped)) != finalDigest).toString(), "true")
}

/** Run [block]; return "ok" if it returns, else the NaalpException kind. */
private fun verifyKind(block: () -> Unit): String = try {
    block()
    "ok"
} catch (e: NaalpException) {
    e.kind
}

fun main() {
    println("streaming conformance (Kotlin) -- graded vs vectors/stream/cases.json")
    streamingRun()
    println(if (stFails == 0) "StreamingTests: PASS" else "StreamingTests: FAIL ($stFails)")
    exitProcess(if (stFails == 0) 0 else 1)
}
